extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```
@@ -0,0 +1,211 @@
+      String recordTimestampLabelPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser; //Header parser can be null by config
+    this.keyParser = keyParser; //Key parser can be null by config and data
+    this.valueParser = valueParser; //value parser can be null by data (tombstone records)
+    this.keyLabelPrefix = keyLabelPrefix;
+    this.recordTimestampLabelPrefix = recordTimestampLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    // Add kafka record timestamp to the mergelist
+    List<Pair<String, Object>> mergeList = new ArrayList<>();
+    if (headerParser != null) {
+      List<Pair<String, Object>> headerList = headerParser.read();
+      mergeList.addAll(headerList);
+    }
+
+    Pair ts = new Pair(recordTimestampLabelPrefix + DEFAULT_TIMESTAMP_STRING,
+                      record.getRecord().timestamp());
+    if (!mergeList.contains(ts)) {
+      mergeList.add(ts);
+    }
+
+    if (keyParser != null) {
+      try (CloseableIterator<InputRow> keyIterator = keyParser.read()) {
+        // Key currently only takes the first row and ignores the rest.
+        if (keyIterator.hasNext()) {
+          // Return type for the key parser should be of type MapBasedInputRow
+          // Parsers returning other types are not compatible currently.
+          MapBasedInputRow keyRow = (MapBasedInputRow) keyIterator.next();
+          Pair key = new Pair(keyLabelPrefix + DEFAULT_KEY_STRING,
+                            keyRow.getEvent().entrySet().stream().findFirst().get().getValue());
+          if (!mergeList.contains(key)) {
+            mergeList.add(key);
+          }
+        }
+      }
+      catch (IOException e) {
+        throw e;
+      }
+      catch (ClassCastException e) {
+        log.error(e, "Encountered ClassCastException exception, please use parsers that return MapBasedInputRow based rows.");
```

no need to explain what exception we encountered, the stacktrace shown in the log would already make that clear.
You'd also want to abort any ingestion rather than logging this message for every single row. To do that throw an IOException so that Druid would treat this like a corrupt input.

to do that I would throw
```suggestion
        throw new IOException("Unsupported input format. KafkaInputformat only supports input format that return MapBasedInputRow rows");
```